### ISSUE TITLE
AllowVMInspection option causes GC pressure / CPU usage

### ISSUE THREAD

**Issue by nhoughto** [oracle/graal#1793]
Playing around with `AllowVMInspection` to allow heapdumps for a simple hello world native image, and I notice that if you enable this option the built native image exhibits quite different behaviour, once enabled the native image constantly allocate and then GC something causing an not insignificant amount of CPU usage / thrashing.

It seems the AllowVMInspection adds quite a lot to a built image (fine) but the ongoing cost of that addition seems much higher than expected, it would be great to be able to enable this option in production / release images so you have the ability to get a heap dump if required (for problems in the wild), but the current behaviour prohibits the flag being used in a release artifact.

Built an image with just this
```
public class TestMain {

    public static void main(String[] args) throws InterruptedException {
        Thread.sleep(60000);
    }
}
```

no vm inspection - 2mb image is built, running it with `-XX:+PrintGC` logs no GC events for ~60 second run.
```
Build on Server(pid: 95126, port: 61842)*
SendBuildRequest [
-task=com.oracle.svm.hosted.NativeImageGeneratorRunner
-H:Name=test-novminspection
-H:CLibraryPath=/home/graal/Contents/Home/jre/lib/svm/clibraries/darwin-amd64
-H:Path=/home/build/graal
-H:Class=TestMain
]
[test-novminspection:95126]    classlist:   5,362.30 ms
[test-novminspection:95126]        (cap):   1,220.91 ms
[test-novminspection:95126]        setup:   1,832.76 ms
[test-novminspection:95126]   (typeflow):   2,061.14 ms
[test-novminspection:95126]    (objects):   1,746.54 ms
[test-novminspection:95126]   (features):     449.70 ms
[test-novminspection:95126]     analysis:   4,306.85 ms
[test-novminspection:95126]     (clinit):      80.34 ms
[test-novminspection:95126]     universe:     206.82 ms
[test-novminspection:95126]      (parse):     115.91 ms
[test-novminspection:95126]     (inline):     538.22 ms
[test-novminspection:95126]    (compile):   1,640.31 ms
[test-novminspection:95126]      compile:   2,434.22 ms
[test-novminspection:95126]        image:     169.83 ms
[test-novminspection:95126]        write:     162.42 ms
[test-novminspection:95126]      [total]:  14,600.27 ms
native image available at build/graal/test-novminspection (2 MB)

BUILD SUCCESSFUL in 18s
16 actionable tasks: 1 executed, 15 up-to-date
~:cli$ ./build/graal/test-novminspection -Xmx64m -XX:+PrintGC -XX:+PrintGCTimeStamps
^C
~:cli$ ./build/graal/test-novminspection -Xmx128m -XX:+PrintGC -XX:+PrintGCTimeStamps
^C
```

With VMInspection - builds a 40mb image, and `-XX:+PrintGC` shows constant collection, every ~1 second (!!)
```
Build on Server(pid: 95126, port: 61842)*
SendBuildRequest [
-task=com.oracle.svm.hosted.NativeImageGeneratorRunner
-H:+AllowVMInspection
-H:Name=test-vminspection
-H:CLibraryPath=/home/graal/Contents/Home/jre/lib/svm/clibraries/darwin-amd64
-H:Path=/home/build/graal
-H:Class=TestMain
]
[test-vminspection:95126]    classlist:   7,311.62 ms
[test-vminspection:95126]        (cap):   2,143.07 ms
[test-vminspection:95126]        setup:   3,894.11 ms
[test-vminspection:95126]   (typeflow):  15,937.98 ms
[test-vminspection:95126]    (objects):  13,663.83 ms
[test-vminspection:95126]   (features):   1,909.08 ms
[test-vminspection:95126]     analysis:  32,880.46 ms
[test-vminspection:95126]     (clinit):     541.85 ms
[test-vminspection:95126]     universe:   1,476.30 ms
[test-vminspection:95126]      (parse):   2,886.48 ms
[test-vminspection:95126]     (inline):   4,074.36 ms
[test-vminspection:95126]    (compile):  51,947.97 ms
[test-vminspection:95126]      compile:  61,398.72 ms
[test-vminspection:95126]        image:   2,834.60 ms
[test-vminspection:95126]        write:     868.33 ms
[test-vminspection:95126]      [total]: 111,091.81 ms
native image available at build/graal/test-vminspection (40 MB)

BUILD SUCCESSFUL in 2m 2s
16 actionable tasks: 7 executed, 9 up-to-date
~:$ ./build/graal/test-vminspection -Xmx64m -XX:+PrintGC -XX:+PrintGCTimeStamps
[213 msec: Incremental GC (CollectOnAllocation.Sometimes) 7139K->1019K, 0.0042773 secs]
[1029 msec: Full GC (CollectOnAllocation.Sometimes) 8159K->1019K, 0.0085179 secs]
[1631 msec: Incremental GC (CollectOnAllocation.Sometimes) 8159K->1019K, 0.0021708 secs]
[2448 msec: Incremental GC (CollectOnAllocation.Sometimes) 8159K->1019K, 0.0039213 secs]
[3059 msec: Full GC (CollectOnAllocation.Sometimes) 8159K->1019K, 0.0080486 secs]
[3866 msec: Incremental GC (CollectOnAllocation.Sometimes) 8159K->1019K, 0.0050320 secs]
[4466 msec: Incremental GC (CollectOnAllocation.Sometimes) 8159K->1019K, 0.0020749 secs]
[5282 msec: Full GC (CollectOnAllocation.Sometimes) 8159K->1019K, 0.0081735 secs]
[5889 msec: Incremental GC (CollectOnAllocation.Sometimes) 8159K->1019K, 0.0047377 secs]
^C
~:$ ./build/graal/test-vminspection -Xmx128m -XX:+PrintGC -XX:+PrintGCTimeStamps
[819 msec: Incremental GC (CollectOnAllocation.Sometimes) 13259K->1019K, 0.0070950 secs]
[2231 msec: Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0043111 secs]
[3450 msec: Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0070408 secs]
[4863 msec: Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0023711 secs]
[6086 msec: Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0041725 secs]
[7509 msec: Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0083310 secs]
[8718 msec: Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0049210 secs]
```

**Comment by thurka** [comment]
Can you, please, tell me, which version of GraalVM do you use?

**Comment by nhoughto** [comment]
Tested this on 19.2.1. Didn’t test older versions.

**Comment by thurka** [comment]
Can you post the output of `/home/graal/Contents/Home/bin/java -version`? Thanks.

**Comment by nhoughto** [comment]
```
java -version
java version "1.8.0_231"
Java(TM) SE Runtime Environment (build 1.8.0_231-b11)
Java HotSpot(TM) 64-Bit GraalVM EE 19.2.1 (build 25.231-b11-jvmci-19.2-b03, mixed mode)
```

**Comment by thurka** [comment]
OK, thanks. `AllowVMInspection` in GraalVM EE enables more that just heapdumps. I would like to see actual CPU overhead for different setups. Can you, please, use `time` to measure CPU usage? Please run (no GC log this time)
`time ./build/graal/test-vminspection -Xmx128m` 
`time ./build/graal/test-vminspection -XX:-UsePerfData -Xmx128m` 
`time ./build/graal/test-novminspection -Xmx128m`
and post the results?

**Comment by nhoughto** [comment]
So 65x more CPU time with VMInspection on, oddly UsePerfData seemed to correct the behaviour? VMInspection w/ UsePerfData behaves the same as no VMInspection

```
time ./build/graal/test-vminspection -Xmx128m -XX:+PrintGC
[Incremental GC (CollectOnAllocation.Sometimes) 13259K->1019K, 0.0070899 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0081391 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0050119 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0066229 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0022257 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0020928 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0079833 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0036563 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0045383 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0084217 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0045627 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0033212 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0075970 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0020332 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0045295 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0078897 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0046682 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0048171 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0083669 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0048359 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0031865 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0074567 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0047302 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0028385 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0082933 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0033459 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0033231 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0079508 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0023245 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0046307 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0052220 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0062633 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0044691 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0078952 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0041835 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0047256 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0068822 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0048795 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0067394 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0033197 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0049409 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0081858 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0044396 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0028916 secs]
[Full GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0052721 secs]
[Incremental GC (CollectOnAllocation.Sometimes) 14279K->1019K, 0.0029586 secs]

real	1m0.051s
user	0m0.260s
sys	0m0.024s

time ./build/graal/test-vminspection -Xmx128m -XX:+PrintGC -XX:-UsePerfData

real	1m0.027s
user	0m0.004s
sys	0m0.008s


time ./build/graal/test-novminspection -Xmx128m -XX:+PrintGC

real	1m0.020s
user	0m0.002s
sys	0m0.004s
```

**Comment by thurka** [comment]
> So 130x more CPU time (with these rough measurements) with VMInspection on, 

Wow, this is really interesting interpretation of the numbers below.

> oddly UsePerfData seemed to correct the behaviour? 

Why is this 'oddly'?

VMInspection w/ UsePerfData behaves the same as no VMInspection
> 
> time ./build/graal/test-vminspection -Xmx128m -XX:+PrintGC

This is not what I asked above - you added `-XX:+PrintGC`, which adds unnecessary CPU usage. Without `-XX:+PrintGC` I see overhead about 0,2% (85 ms).
```
hugo:19.2.1 thurka$ time ./testmain -Xmx128m

real	1m0.023s
user	0m0.002s
sys	0m0.004s
hugo:19.2.1 thurka$ time ./testmain_vminspection -Xmx128m

real	1m0.018s
user	0m0.073s
sys	0m0.018s
```

**Comment by thurka** [comment]
> With VMInspection - builds a 40mb image.

I don't see 40mb image with VMInspection enabled. My executable from GraalVM EE 19.2.1 is 6.6M.
```
hugo:19.2.1 thurka$ ls -l
total 18208
-rw-r--r--  1 thurka  wheel      386 Oct 31 10:52 TestMain.class
-rw-r--r--  1 thurka  wheel      134 Oct 31 10:52 TestMain.java
-rwxr-xr-x  1 thurka  wheel  2432144 Oct 31 10:55 testmain
-rwxr-xr-x  1 thurka  wheel  6881252 Oct 31 10:54 testmain_vminspection
```

**Comment by nhoughto** [comment]
Will run again without printgc.

It’s odd because I have no idea what “usePerfData” does, and it seems strange that one flag would exist that would exclude another without any error or warning 🤷🏻‍♂️. To me that seems odd!

**Comment by nhoughto** [comment]
so 21x without PrintGC

```
time ./test-novminspection 

real	1m0.029s
user	0m0.003s
sys	0m0.006s

time ./test-vminspection 

real	1m0.057s
user	0m0.065s
sys	0m0.038s


time ./test-vminspection -XX:-UsePerfData

real	1m0.015s
user	0m0.004s
sys	0m0.006s
```

**Comment by nhoughto** [comment]
So just rounding this out, if you want to use vmsinpection with no overhead you must add `-UsePerfData` on command line. Great to have a workaround, but since you can't bake cli arguments into the built native image (yet?) it would require all consumer of the image to add this argument which isn't great ergonmics.

**Comment by christianhaeubl** [comment]
Fixed in 20.3.
